Support Container References in `WithImage`

afscrome opened the pull request:
## Description

Updated `WithImage()` to properly handle full image references, parsing out out registry, tag & images.  This means you can use any string you pass to `docker run` for an image e.g. `WithImage("mcr.microsoft.com/dotnet/aspire-dashboard:8.2")`.

Fixes #4355
Fixes #5834


## Checklist

- Is this feature complete?
  - [X] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [X] Yes
  - [ ] No
- Did you add public API?
  - [X] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [X] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [X] Yes
      - [ ] No
  - [ ] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [X] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [X] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5835)

afscrome committed:
Add container reference parser

davidfowl commented on src/Aspire.Hosting/Utils/ContainerReferenceParser.cs:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    // Based on https://github.com/distribution/reference/blob/8c942b0459dfdcc5b6685581dd0a5a470f615bff/regexp.go
+    public class ContainerParserRegexes
```
this should be using the regex source generator

afscrome committed:
Merge remote-tracking branch 'upstream/main' into withimagereference

afscrome committed:
Fixes based on latest discussions.

afscrome commented on src/Aspire.Hosting/Utils/ContainerReferenceParser.cs:
```diff
@@ -0,0 +1,43 @@
+
+        if (!match.Success)
+        {
+            throw new ArgumentOutOfRangeException(nameof(input), input, "invalid reference format");
```
The error message isn't great, but it matches the error docker gives off, so 
https://github.com/distribution/reference/blob/8c942b0459dfdcc5b6685581dd0a5a470f615bff/reference.go#L72

danegsta replied:
Maybe something like "invalid reference format: could not parse container image name"?

afscrome commented on src/Aspire.Hosting/Utils/ContainerReferenceParser.cs:
```diff
@@ -0,0 +1,43 @@
+            => group.Success ? group.Value : default;
+    }
+
+    [GeneratedRegex("^((?<registry>((localhost|\\[(?:[a-fA-F0-9:]+)\\]|([\\w-]+(\\.[\\w-]+)+))(:\\d+)?)|([\\w-]+:\\d+))\\/)?(?<image>[\\w-./<>]+)(:(?<tag>[\\w-.<>]+))?(@(?<digest>.+))?$")]
```
Based on https://github.com/dotnet/aspire/blob/ea985ec4fd70b9eb8cb2f8c1a567372f5475c89c/src/Aspire.Hosting/Utils/ContainerReferenceParser.cs#L40 with addition for IPv6 addresses

afscrome commented on src/Aspire.Hosting/ContainerResourceBuilderExtensions.cs:
```diff
@@ -138,18 +139,57 @@ public static IResourceBuilder<T> WithImageRegistry<T>(this IResourceBuilder<T>
+
+        if (imageAnnotation.Tag is null && imageAnnotation.SHA256 is null)
+        {
+            imageAnnotation.Tag = "latest";
```
This bit could instead be done by defaulting the tag in `ContainerImageAnnotation`

afscrome commented on src/Aspire.Hosting/ContainerResourceBuilderExtensions.cs:
```diff
@@ -138,18 +139,57 @@ public static IResourceBuilder<T> WithImageRegistry<T>(this IResourceBuilder<T>
     /// <param name="tag">Tag value.</param>
     /// <returns></returns>
-    public static IResourceBuilder<T> WithImage<T>(this IResourceBuilder<T> builder, string image, string tag = "latest") where T : ContainerResource
+    public static IResourceBuilder<T> WithImage<T>(this IResourceBuilder<T> builder, string image, string? tag = null) where T : ContainerResource
```
I've had to make an API breaking change here - specifically changing `string tag = "latest"` to `string? tag = null`.

Without this, `WithImage("foo:tag")` would become `WithImage("foo:tag", "latest")` which would be ambiguous.

Open to other suggestions

mitchdenny replied:
If we are going to break it might be better to break it by removing the tag argument altogether.

Thoughts @davidfowl / @danegsta ?

afscrome replied:
Would splitting the method up into two be a better option than changing the default value?
```cs
public static IResourceBuilder<T> WithImage<T>(this IResourceBuilder<T> builder, string image) where T : ContainerResource

// Possibly with addition of [Obsolete("Use 'WithImage(image).WithTag(tag) instead.")]
public static IResourceBuilder<T> WithImage<T>(this IResourceBuilder<T> builder, string image, string tag) where T : ContainerResource
    => builder.WithImage(image).WithImageTag(tag);
```

For my understanding, what exactly is the breaking change the public api analyser is complaining about here?  Is it the `string!` to `string?` - I'd have thought that would be OK as we're widening allowed inputs, not narrowing or is it the change of the default value itself.  Or is it the change of the default value from `"latest"` to `null`?  I know changing the default value is problematic for binary compatibility as anyone who previously called the method will still have the old `"latest"` default baked in, however in this particular case we ultimately set the tag to `"latest"` if the tag is `null` so the impact of changing the default should be mostly non observable.

afscrome replied:
Any thoughts @davidfowl / @danegsta ?

davidfowl replied:
I like the idea of making the default tag null and we’d internally still use latest only if the tag wasn’t set. Feels like we should make this change now if we’re going to do it

afscrome replied:
@davidfowl  Just to check, do you mean handle the default value within the `WithImage` method as done in the PR (i.e. setting ContainerImageAnnotation.Tag to `latest` method if the tag argument is null / missing), or do you mean handle the default elsewhere (i.e. leave `ContainerImageAnnotation.Tag` as null and then handling that later on.)

davidfowl replied:
probably as delayed as possible. Though I am concered at the number of places that might need to update the handling of a null tag.

afscrome replied:
3 types of tests broke

1. Tests asserting that `ContainerImageAnnotation.Tag` was `latest` when they didn't pass a tag. (expected to break as that's what we changed)
2. Mainfests for images with tags no longer have `:latest` in theim  (i.e. `nginx:latest` has become just `nginx`)
3. Image in bicep files without a tag no longer have `:latest` on them.

Both 2 & 3 share the same root cause with `TryGetContainerImageName` which currently has explicitly logic to not include the tag if one isn't set  https://github.com/dotnet/aspire/blob/0b562b8eb039343df763d6e9083da8fbb938604e/src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs#L242-L246

This change also removes the `:latest` from the dashboard UI if no tag is specified:

![image](https://github.com/user-attachments/assets/6ac79ce0-6fff-4bd7-bc83-0be96f0754be)

```cs
builder.AddContainer("no-tag", "nginx");
builder.AddContainer("explicit-latest", "nginx:latest");
```

So the question is should aspire explicitly set the tag to latest before passing to the manifest / container runtime, or should it just leave the tag blank and let that system deal with it?  The PR currently has the latter, but changing the former should be a matter of changing `TryGetContainerImageName`, backing out the test changes from b93a3f2f59c8e0804e887222576e9ab32f3a6f26 and 81a77b3b245a6cffe16d778af6608be6a0ce0f95 and adding some additional test coverage.

afscrome committed:
Fix tests

- If an existing registry was set, and `WithImage()` was not given a registry, keep the existing registry to match tag & digest behaviour.
- Removed rabbitmq test case which now throws earier error as " " is an invalid image name.

afscrome commented:
@danegsta @mitchdenny Do you mind taking a look at this?  The two test failures seem to be known issues - #6163

mitchdenny commented on tests/Aspire.Hosting.RabbitMQ.Tests/AddRabbitMQTests.cs:
```diff
@@ -163,7 +163,6 @@ public void WithManagementPluginThrowsForUnsupportedContainerImageTag(string ima
     }
 
     [Theory]
-    [InlineData(" ")]
```
Why this change?

afscrome replied:
`rabbitmq.WithImage(" ")` now throws `invalid reference format` before the test can even get as far as testing if `rabbitmq.WithManagementPlugin` throws.

This scenario is covered by https://github.com/dotnet/aspire/blob/03b07232818f172246dbc274b3fc1fa6a9a07cb6/tests/Aspire.Hosting.Tests/Utils/ContainerReferenceParserTests.cs#L20

afscrome replied:
@mitchdenny does my explanation resolve your question about the test change?

afscrome committed:
Clarify error when image fails to parse.

afscrome committed:
Merge remote-tracking branch 'upstream/main' into withimagereference

afscrome committed:
Don't default tag to "latest" if one isn't provided.

davidfowl commented on tests/Aspire.Hosting.Containers.Tests/ContainerResourceTests.cs:
```diff
@@ -86,7 +86,7 @@ public async Task AddContainerWithArgs()
         {
           "type": "container.v0",
-          "image": "none:latest",
+          "image": "none",
```
This seems bad

afscrome replied:
If a user explicitly gives us a tag of `latest`, then we should include that in the manifest (i.e. `none:latest`).  If a user doesn't' give us a tag that's more open question.  My take was that the images `none` and `none:latest` are equivalent, so if a user doesn't give us a tag we should treat that by not explicitly including a tag.  Happy to put `latest` back in the manifest if you think that's better.

That said, I haven't put huge amounts of thought into this latest commit - mainly done the obvious change to see what tests break.

afscrome committed:
More occurences of latest

davidfowl commented:
@afscrome if we want to take another swing at this, I think we should go with the new method, non breaking approach.

afscrome commented:
Superseded by #7254 7254
